app/vminsert/netstorage/insert_ctx.go

```diff
@@ -165,9 +189,24 @@ func (ctx *InsertCtx) GetStorageNodeIdx(at *auth.Token, labels []prompb.Label) i
 	}
 	h := xxhash.Sum64(buf)
```

Comment:
It is better from performance PoV creating a separate function for calculating this hash only once before looping over all the replication groups and writing data to them.